nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/FetchElasticsearchHttp.java

```diff
@@ -0,0 +1,293 @@
+        }
+
+        OkHttpClient okHttpClient = getClient();
+
+        if (flowFile == null) {
+            flowFile = session.create();
+        }
+
+        final String index = context.getProperty(INDEX).evaluateAttributeExpressions(flowFile).getValue();
+        final String docId = context.getProperty(DOC_ID).evaluateAttributeExpressions(flowFile).getValue();
+        final String docType = context.getProperty(TYPE).evaluateAttributeExpressions(flowFile).getValue();
+        final String fields = context.getProperty(FIELDS).isSet()
+                ? context.getProperty(FIELDS).evaluateAttributeExpressions(flowFile).getValue()
+                : null;
+
+        final ComponentLog logger = getLogger();
+
+
+        try {
+            logger.debug("Fetching {}/{}/{} from Elasticsearch", new Object[]{index, docType, docId});
+
+            // read the url property from the context
+            final String urlstr = StringUtils.trimToEmpty(context.getProperty(ES_URL).getValue());
+            final URL url = buildRequestURL(urlstr, docId, index, docType, fields);
+            final long startNanos = System.nanoTime();
+
+            final Response getResponse = sendRequestToElasticsearch(okHttpClient, url, "GET", null);
+            final int statusCode = getResponse.code();
+
+            if (isSuccess(statusCode)) {
+                ResponseBody body = getResponse.body();
+                final byte[] bodyBytes = body.bytes();
+                JsonNode responseJson = parseJsonResponse(new ByteArrayInputStream(bodyBytes));
+                boolean found = responseJson.get("found").asBoolean(false);
+
+                if (found) {
+                    flowFile = session.putAttribute(flowFile, "filename", docId);
+                    flowFile = session.putAttribute(flowFile, "es.index", index);
+                    flowFile = session.putAttribute(flowFile, "es.type", docType);
+                    flowFile = session.write(flowFile, out -> {
+                        out.write(bodyBytes);
+                    });
+                    logger.debug("Elasticsearch document " + docId + " fetched, routing to success");
+
+                    // emit provenance event
+                    final long millis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+                    session.getProvenanceReporter().fetch(flowFile, url.toExternalForm(), millis);
+                    session.transfer(flowFile, REL_SUCCESS);
+                } else {
+                    logger.warn("Failed to read {}/{}/{} from Elasticsearch: Document not found",
+                            new Object[]{index, docType, docId});
+
+                    // We couldn't find the document, so penalize it and send it to "not found"
+                    // If there was an incoming flow file, penalize and transfer. Otherwise just remove it
+                    if (context.hasIncomingConnection()) {
+                        flowFile = session.penalize(flowFile);
```

Comment:
Best practice is to penalize a flowfile when the issue the flowfile ran into may be resolved if the flowfile was re-processed later (like for REL_RETRY) but for this relationship the document was not found at all and there is no guarantee, if it was run again later, that it would succeed. It would probably be best not to penalize the flowfile here.